elastic/kibana
# [Monitoring] Server side pagination for ES Nodes listing table

Relates to #36358

This PR introduces server-side pagination for the Elasticsearch nodes listing page

See #46587 for more specifics, as the same approach was taken here.

One deviation is I needed to add a high-level terms query around the `getMetrics` call done in order to properly associated each set of aggregated data with the proper node. To support this, I added logic to add a `groupBy` when fetching metrics that would result in a high-level `terms` query, and the response would be nearly the same, but each metric will have a `groupedBy` field with the key from the terms aggregation result.

## Performance
I wanted to test the performance on a real-life use case, preferably one where the nodes page takes awhile to load to right now. 

I found an ESMS cluster that had 59 nodes.

Using this ESMS cluster, I configured two separate Kibana instances to point to it for monitoring data - one running master and one running this PR. I used curl (and the `time_starttransfer` data point) to query the `/api/monitoring/v1/clusters/<cluster_uuid>/elasticsearch/nodes` endpoint and time the response with a sample size of 50 requests. For this PR, I sent this for pagination: `{index:0,size:10}` and the time range is the same `1h` period of time. One important thing to note is the `search.max_buckets` for this cluster is set to `50,000`.

### Response Time
Master: `1.8111`
PR: `0.6380`

This is an improvement of **65%** in terms of response time.

### Buckets created
Master: `7198`
PR: `1220`

That's an improvement of **83%** in terms of buckets created.

This is pretty significant and should help out many users loading and using these two pages.

## Testing
- Start a cluster with multiple ES nodes
- View the ES nodes listing pages and ensure the pagination/sorting/filtering works as expected.

## TODO
- [x] ~~Add api integration tests~~
- [x] ~~Add setup mode support to `eui_table_ssp.js`~~
